pkg/internal/syncs/syncs.go

```diff
@@ -0,0 +1,37 @@
+package syncs
+
+import (
+	"context"
+	"reflect"
+	"sync"
+)
+
+// MergeChans returns a channel that is closed when any of the input channels are signaled.
+// The caller must call the returned CancelFunc to ensure no resources are leaked.
+func MergeChans[T any](chans ...<-chan T) (<-chan T, context.CancelFunc) {
+	var once sync.Once
+	out := make(chan T)
+	cancel := make(chan T)
+	cancelFunc := func() {
+		once.Do(func() {
+			close(cancel)
```

Comment:
I think we have to wait for out to be canceled here.
Currently `cancel()` can complete without `chan out` being closed.

Reply:
Are you saying that you'd like to see `cancelFunc()` block until `out` is definitely closed?

This is different than how context.WithCancel() [works](https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/context/context.go;l=471), but I don't see that as proof-positive it shouldn't be done how you're suggesting.

What is the use case for blocking-until-closed? That won't necessarily guarantee that any goroutines that are expected to respond to `out` have done so (e.g. maybe they are running an infinite for-loop, occasionally checking to see if `out` is closed). To guarantee that, each goroutine would need to send a `done` signal, confirming they have shut down.

Or do you see a bug where `out` may not successfully be closed?

Reply:
I'm not sure what an example of code that relies on cancel blocking on channel close would look like? Generally cancel() signals child goroutines that they should exit.

To avoid arguing, I've added it in. It does mean that future coding errors have a higher chance of causing deadlocks, but I think is fine as-written.